otelcol/collector.go

```diff
@@ -109,14 +110,62 @@ type Collector struct {
 	signalsChannel chan os.Signal
 	// asyncErrorChannel is used to signal a fatal error from any component.
 	asyncErrorChannel chan error
+	cc                *collectorCore
+}
+
+var _ zapcore.Core = &collectorCore{}
+
+type collectorCore struct {
+	core zapcore.Core
+}
+
+func (c *collectorCore) Enabled(l zapcore.Level) bool {
+	return c.core.Enabled(l)
+}
+
+func (c *collectorCore) With(f []zapcore.Field) zapcore.Core {
+	return c.core.With(f)
+}
+
+func (c *collectorCore) Check(e zapcore.Entry, ce *zapcore.CheckedEntry) *zapcore.CheckedEntry {
+	return c.core.Check(e, ce)
+}
+
+func (c *collectorCore) Write(e zapcore.Entry, f []zapcore.Field) error {
+	return c.core.Write(e, f)
+}
+
+func (c *collectorCore) Sync() error {
+	return c.core.Sync()
+}
+
+func newCollectorCore(options []zap.Option) (*collectorCore, error) {
+	ec := zap.NewProductionEncoderConfig()
+	ec.EncodeTime = zapcore.ISO8601TimeEncoder
+	zapCfg := &zap.Config{
+		Level:            zap.NewAtomicLevelAt(zapcore.InfoLevel),
+		Encoding:         "console",
+		EncoderConfig:    ec,
+		OutputPaths:      []string{"stderr"},
+		ErrorOutputPaths: []string{"stderr"},
+	}
+	logger, err := zapCfg.Build(options...)
+	if err != nil {
+		return nil, err
+	}
+	return &collectorCore{core: logger.Core()}, nil
```

Comment:
This was my lazy way of getting a useable core to wrap, there is probably a better way

Reply:
This seems to work, in the actual implementation I guess `newCollectorCore` would start with an observer and also return some sort of handler to update the core once the logger is ready. We would need to handle concurrent access to the `core` field inseide `collectorCore` as well

Reply:
Ya I could combine this with the `observer` stuff.  I'll touch it up.

Reply:
Updated.  Here is what the Warning log looks like now:

![image](https://github.com/open-telemetry/opentelemetry-collector/assets/12352919/640a3ea5-60ee-4de0-b5dd-4b41a984c927)

Reply:
Personally I still like the idea of passing a basic logger to confmap to handle logging before the config has been rresolved, and then updating the confmap's logger to the configured logger once we're able.

Reply:
It feels like this would solve most of the problems. I guess the problem is that someone's ever going to use a configuration to emit logs via OTLP directly, they may never see logs for a configuration failure. I don't know if there's a sane way to handle this other than via documentation though, since without config you can't configure the logger you wanted to emit logs with.

Reply:
I am going to write up a rfc doc for how the collector should handle logging before it has a logger instantiated, that should help us clarify the behavior we want to implement in this PR.